test/unit/org/apache/cassandra/transport/NativeTransportTimeoutTest.java

```diff
@@ -0,0 +1,170 @@
+                                "flag(Thread.currentThread());"),
+                       @BMRule(name = "Mock NTR timeout from handleRequest",
+                       targetClass = "org.apache.cassandra.config.DatabaseDescriptor",
+                       targetMethod = "getNativeTransportTimeout",
+                       targetLocation = "AT ENTRY",
+                       condition = "flagged(Thread.currentThread()) && callerEquals(\"handleRequest\")",
+                       action = "clear(Thread.currentThread()); " +
+                                "return 10000000;") })
+    public void testAsyncStageLoadShedding() throws Throwable
+    {
+        CassandraRelevantProperties.NATIVE_TRANSPORT_ASYNC_READ_WRITE_ENABLED.setBoolean(true);
+
+        try
+        {
+            createTable("CREATE TABLE %s (pk int PRIMARY KEY, v text)");
+
+            Statement statement = new SimpleStatement("SELECT * FROM " + KEYSPACE + '.' + currentTable());
+            doTestLoadShedding(true, statement);
+
+            Statement insert1 = new SimpleStatement("INSERT INTO " + KEYSPACE + '.' + currentTable() + " (pk, v) VALUES (1, 'foo')");
+            Statement insert2 = new SimpleStatement("INSERT INTO " + KEYSPACE + '.' + currentTable() + " (pk, v) VALUES (2, 'bar')");
+            statement = new BatchStatement().add(insert1).add(insert2);
+            doTestLoadShedding(true, statement);
+
+            PreparedStatement ps = sessionNet().prepare("SELECT * FROM " + KEYSPACE + '.' + currentTable());
+            doTestLoadShedding(true, ps.bind());
+        }
+        finally
+        {
+            CassandraRelevantProperties.NATIVE_TRANSPORT_ASYNC_READ_WRITE_ENABLED.setBoolean(false);
+        }
+    }
+
+    private void doTestLoadShedding(boolean useAsyncStages, Statement statement) throws InterruptedException
+    {
+        EXECUTE_BARRIER = new Semaphore(0);
+        WAIT_BARRIER = new Semaphore(0);
+
+        Meter timedOutMeter;
+        Timer queueTimer;
+
+        Session session = sessionNet();
+
+        // custom payload used to make detection of this statement easy early in byteman rules
+        statement.setOutgoingPayload(Collections.singletonMap("sentinel", ByteBuffer.wrap(new byte[0])));
+
+        if (useAsyncStages)
+        {
+            timedOutMeter = ClientMetrics.instance.timedOutBeforeAsyncProcessing;
+            queueTimer = ClientMetrics.instance.asyncQueueTime;
+        }
+        else
+        {
+            timedOutMeter = ClientMetrics.instance.timedOutBeforeProcessing;
+            queueTimer = ClientMetrics.instance.queueTime;
+        }
+
+        long initialTimedOut = timedOutMeter.getCount();
+
+        ResultSetFuture rsf = session.executeAsync(statement);
+        WAIT_BARRIER.acquire();
+        Thread.sleep(10 + TimeUnit.MILLISECONDS.convert(approxTime.error(), TimeUnit.NANOSECONDS) * 2);
```

Comment:
it would be nice to have a comment here explaining why 10 (pointing to byteman rules)